### PR TITLE
v4.0.0 Alpha Feature Additions

### PR DESCRIPTION
**Pull request by retiutut**
Address the following issues:

#48 
#55 
and full integration with the solution for #38 

**Add Feature: Playback Widget**
- [x] Time Series update the number of channels when loading successive files using playback widget

- [x] EMG update the number of channels when loading successive files using playback widget 
https://github.com/OpenBCI/OpenBCI_GUI/pull/383/commits/d6bf66dc98b6b914a1ec0dee966da9e44271cb11

- [x] Headplot update the number of channels when loading successive files using playback widget
https://github.com/OpenBCI/OpenBCI_GUI/pull/383/commits/8ac64e18d9206ae28f8aa56ce3a9d12699b865d5

- [x] Add feature to fully address #55 and make a history of playback files show within the playback widget, use JSON to make a history file for 10 entries

- [x] Figure out if 100 history entries can fit in the playback widget using a vertical or horizontal scrollbar

- [x] Dropdown menu to select groups of 10 playback files 
https://github.com/OpenBCI/OpenBCI_GUI/pull/383/commits/b74f99f58c9390cab96fecb93c626c9aa49c4bf3

- [x] add N buttons to main window of playback widget allowing user to load the corresponding playback file, where N is the number of files displayed in the window 1-10

- [x] Update Changelog to reflect changes made in this pull request 
https://github.com/OpenBCI/OpenBCI_GUI/pull/383/commits/457d4a8287a36ec353e2e9105c7d22e8df44c9d1

**Add Feature: Check latest version button** 
https://github.com/OpenBCI/OpenBCI_GUI/pull/383/commits/70d17f0db82ebb0cc8690ea308e89cb91ce96e19 
https://github.com/OpenBCI/OpenBCI_GUI/pull/383/commits/f0f21197b0a9959f873971d1e6b6115a594db1bb
https://github.com/OpenBCI/OpenBCI_GUI/pull/383/commits/4c3cec1181a3bd629bff29a24184a50532ffdb01

- [x] Fetch and parse the latest version from Github using simple code

- [x] create a button

- [x] perform a version check

- [x] allow option to open web browser to latest version page

### CODE REVIEW

**Comment by andrewjaykeller** [comment]
What’s happening here lol

**Comment by retiutut** [comment]
Good things! Not ready to merge right now.

**Comment by retiutut** [comment]
screen shot 2018-09-23 at 6 14 21 pm

**Comment by retiutut** [comment]
screen shot 2018-09-25 at 1 49 13 pm

**Comment by retiutut** [comment]
@andrewjaykeller Ready for review!

**Comment by retiutut** [comment]
screen shot 2018-09-26 at 9 02 21 pm

File: OpenBCI_GUI/ControlPanel.pde [hunk 1]
```diff
@@ -135,9 +135,9 @@ Button synthChanButton4;
 Button synthChanButton8;
 Button synthChanButton16;
 
-Button playbackChanButton4;
-Button playbackChanButton8;
-Button playbackChanButton16;
+//Button playbackChanButton4;
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: remove commented out code

Comment (andrewjaykeller) [reply OpenBCI_GUI/ControlPanel.pde]: it's okay to remove unused code! we have github, we can always go back in time.

File: OpenBCI_GUI/ControlPanel.pde [hunk 2]
```diff
@@ -217,10 +217,10 @@ public void controlEvent(ControlEvent theEvent) {
       wifiIPAddressDyanmic.color_notPressed = isSelected_color;
       wifiIPAddressStatic.color_notPressed = autoFileName.color_notPressed;
     } else if(newDataSource == DATASOURCE_PLAYBACKFILE){
-      updateToNChan(8);
-      playbackChanButton4.color_notPressed = autoFileName.color_notPressed;
-      playbackChanButton8.color_notPressed = isSelected_color;
-      playbackChanButton16.color_notPressed = autoFileName.color_notPressed;
+      updateToNChan(8); //default to 8 chan before user selects a playback file
+      //playbackChanButton4.color_notPressed = autoFileName.color_notPressed;
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code

File: OpenBCI_GUI/ControlPanel.pde [hunk 3]
```diff
@@ -297,7 +297,7 @@ class ControlPanel {
   ChannelCountBox channelCountBox;
   InitBox initBox;
   SyntheticChannelCountBox synthChannelCountBox;
-  PlaybackChannelCountBox playbackChannelCountBox;
+  //PlaybackChannelCountBox playbackChannelCountBox;
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code

File: OpenBCI_GUI/ControlPanel.pde [hunk 4]
```diff
@@ -371,8 +371,8 @@ class ControlPanel {
     wifiTransferProtcolCytonBox = new WifiTransferProtcolCytonBox(x + w + x + w - 3, (latencyCytonBox.y + latencyCytonBox.h), w, h, globalPadding);
 
     //boxes active when eegDataSource = Playback
-    playbackChannelCountBox = new PlaybackChannelCountBox(x + w, dataSourceBox.y, w, h, globalPadding);
-    playbackFileBox = new PlaybackFileBox(x + w, (playbackChannelCountBox.y + playbackChannelCountBox.h), w, h, globalPadding);
+    //playbackChannelCountBox = new PlaybackChannelCountBox(x + w, dataSourceBox.y, w, h, globalPadding);
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code

File: OpenBCI_GUI/ControlPanel.pde [hunk 5]
```diff
@@ -435,7 +435,7 @@ class ControlPanel {
     dataLogBox.update();
     channelCountBox.update();
     synthChannelCountBox.update();
-    playbackChannelCountBox.update();
+    //playbackChannelCountBox.update();
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code

File: OpenBCI_GUI/ControlPanel.pde [hunk 6]
```diff
@@ -595,7 +595,7 @@ class ControlPanel {
           cp5.get(MenuList.class, "sdTimes").setVisible(true); //make sure the SD time record options menulist is visible
         }
       } else if (eegDataSource == DATASOURCE_PLAYBACKFILE) { //when data source is from playback file
-        playbackChannelCountBox.draw();
+        //playbackChannelCountBox.draw();
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code

File: OpenBCI_GUI/ControlPanel.pde [hunk 7]
```diff
@@ -1087,6 +1087,7 @@ class ControlPanel {
           selectSDFile.wasPressed = true;
         }
 
+        /*
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code block

File: OpenBCI_GUI/ControlPanel.pde [hunk 8]
```diff
@@ -1446,6 +1448,7 @@ class ControlPanel {
       cyton.setSampleRate(1000);
     }
 
+    /*
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code block

File: OpenBCI_GUI/ControlPanel.pde [hunk 9]
```diff
@@ -1606,12 +1610,14 @@ class ControlPanel {
     synthChanButton8.wasPressed = false;
     synthChanButton16.setIsActive(false);
     synthChanButton16.wasPressed = false;
+    /*
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code block

File: OpenBCI_GUI/ControlPanel.pde [hunk 10]
```diff
@@ -2606,6 +2612,8 @@ class SyntheticChannelCountBox {
   }
 };
 
+//This has been removed since number of channels is auto-set when file is loaded
+/*
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/ControlPanel.pde]: please remove unused code block

File: OpenBCI_GUI/TopNav.pde [hunk 11]
```diff
@@ -39,6 +43,13 @@ class TopNav {
 
   boolean finishedInit = false;
 
+  String webGUIVersionString;
+  int webGUIVersionInt;
+  int localGUIVersionInt;
+  Boolean guiVersionIsUpToDate;
+  Boolean unableToFindPlistFile;
+
+
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/TopNav.pde]: remove extra lines

File: OpenBCI_GUI/W_EMG.pde [hunk 12]
```diff
@@ -246,6 +226,28 @@ class W_emg extends Widget {
...
+      }
+    }
+
+    events = new boolean[_nchan];
+
+    for (int i = 0; i < _nchan; i++) {
+      events[i] = true;
+    }
+  }
+
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/W_EMG.pde]: remove extra lines

File: OpenBCI_GUI/DataLogging.pde [hunk 13]
```diff
@@ -79,14 +79,54 @@ void openNewLogFileODF(String _fileName) {
...
+      updateToNChan(4);
+      break;
+    case (totalColumnsEightChannels):
+      updateToNChan(8);
+      break;
+    case (totalColumnsSixteenChannels):
+      updateToNChan(16);
+      break;
   }
+  //outputSuccess(str(datatable.getColumnCount()));
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/DataLogging.pde]: remove line

File: OpenBCI_GUI/OpenBCI_GUI.pde [hunk 14]
```diff
@@ -246,8 +246,12 @@ ButtonHelpText buttonHelpText;
 boolean no_start_connection = false;
 boolean has_processed = false;
 boolean isOldData = false;
-
+//Used for playback file
 int indices = 0;
+//# columns used by a playback file determines number of channels
+final int totalColumnsFourChannels = 8;
+final int totalColumnsEightChannels = 12;
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/OpenBCI_GUI.pde]: what if someone uses analog data and changes the number of columns, or timestamp changes the number of columns? we should rethink using a hard count.

File: OpenBCI_GUI/OpenBCI_GUI.pde [hunk 15]
```diff
@@ -345,6 +349,25 @@ Boolean errorUserSettingsNotFound = false; //For error catching
 int loadErrorTimerStart;
 int loadErrorTimeWindow = 5000; //Time window in milliseconds to apply channel settings to Cyton board. This is to avoid a GUI crash at ~ 4500-5000 milliseconds.
 Boolean loadErrorCytonEvent = false;
+Boolean settingsLoadedCheck = false; //Used to determine if settings are done loading successfully after init
+
+//Used to check GUI version in TopNav.pde and displayedon the splash screen on startup
+String localGUIVersionString = "v4.0.0-alpha.0";
```
Comment (andrewjaykeller) [inline OpenBCI_GUI/OpenBCI_GUI.pde]: update this to `alpha.2`

**Comment by andrewjaykeller** [comment]
I think the biggest worry i have is if you record digital inputs with the cyton, for example, the number of columns would change recorded to file.

File: OpenBCI_GUI/OpenBCI_GUI.pde (hunk 14 above)
Comment (retiutut) [reply OpenBCI_GUI/OpenBCI_GUI.pde]: I agree.

Comment (retiutut) [reply OpenBCI_GUI/OpenBCI_GUI.pde]: @andrewjaykeller So use a range of channels? Or just count the number of columns after the index only?

**Pull request merged**